packages/yswaps/scripts/libraries/solvers/multicall/CurveSpellEth.ts

```diff
@@ -88,41 +101,55 @@ export class CurveSpellEth implements Solver {
       slippagePercentage: 10 / 100,
     });
 
-    const approveCvx = (await crv.allowance(multicallSwapperAddress, zrxCvxAllowanceTarget)) < cvxBalance;
-    if (approveCvx) await cvx.approve(zrxCvxAllowanceTarget, constants.MaxUint256);
+    const approveCvx = (await cvx.allowance(multicallSwapperAddress, zrxCvxAllowanceTarget)).lt(cvxBalance);
+    if (approveCvx) {
+      console.log('[CurveSpellEth] Approving cvx');
+      const approveCvxTx = await cvx.populateTransaction.approve(zrxCvxAllowanceTarget, constants.MaxUint256);
+      await multicallSwapperSigner.sendTransaction(approveCvxTx);
+      transactions.push(approveCvxTx);
+    }
 
+    console.log('[CurveSpellEth] Executing cvx => weth via zrx');
     const cvxToWethTx: Tx = {
       to: this.zrxContractAddress,
       data: zrxCvxData,
     };
     await multicallSwapperSigner.sendTransaction(cvxToWethTx);
+    transactions.push(cvxToWethTx);
 
-    console.log('[CurveSpellEth] Convert weth to eth');
     const wethBalance = await weth.balanceOf(multicallSwapperAddress);
-    await weth.withdraw(wethBalance);
-
-    console.log('[CurveSpellEth] Convert weth to crvSpellEth');
-    await curveSwap.add_liquidity([wethBalance, 0], 0, true, this.strategyAddress, { value: wethBalance });
+    console.log('[CurveSpellEth] Total WETH balance is', utils.formatEther(wethBalance));
+
+    const approveWeth = (await weth.allowance(multicallSwapperAddress, multicallSwapperAddress)).lt(wethBalance);
+    if (approveWeth) {
+      console.log('[CurveSpellEth] Approving weth');
+      const approveWethTx = await weth.populateTransaction.approve(curveSwap.address, constants.MaxUint256);
+      await multicallSwapperSigner.sendTransaction(approveWethTx);
+      transactions.push(approveWethTx);
+    }
+
+    console.log('[CurveSpellEth] Converting eth to crvSpellEth');
+    const curveCalculatedTokenAmountOut = await curveSwap.calc_token_amount([wethBalance, 0]);
+    const curveCalculatedTokenMinAmountOut = curveCalculatedTokenAmountOut.sub(curveCalculatedTokenAmountOut.mul(3).div(100)); // 3% slippage
+    const addLiquidityTx = await curveSwap.populateTransaction.add_liquidity(
+      [wethBalance, 0],
+      curveCalculatedTokenMinAmountOut,
+      false,
+      this.strategyAddress
```

Comment:
should we keep using `this.strategyAddress` in few places since now we are using function's param `strategy` in replace in several places?

Reply:
Honestly ? We shouldn't, because the matching will be done based in the configuration (where the strat is the key of the map). We might leave an array here of strats that are able to use the solver, just to double-check, be safe. But no, this is not needed. You can create a refactor branch once we merge this one and take 'em out :)